### ISSUE TITLE
Python 3.13 support

### ISSUE THREAD

**Issue by SnoopJ** [sopel-irc/sopel#2608]
### Requested Feature

Sopel should support Python 3.13 when possible. 3.13 entered beta on 7 May and is now feature complete, so it's a good time to start tracking what's needed for support. 3.13.0 final is scheduled for October 2024, and release candidates are scheduled for late July and early September.

### Notes

Known problems:

* ~required `setuptools` version needs to be updated to 66.0 or higher (usage of deprecated `pkgutil.ImpImporter` by `setuptools.version`)~ done in #2516
* ~`greenlet` (required by `SQLAlchemy`) does not yet support Python 3.13 ([issue](https://github.com/python-greenlet/greenlet/issues/392))~
  * ~There _is_ a PR to add support, but I can't build it against 3.13.0b1: https://github.com/python-greenlet/greenlet/pull/396~
  * Fixed in `greenlet` 3.1.0: https://github.com/python-greenlet/greenlet/pull/396

**Comment by SnoopJ** [comment]
It may merit a new issue, but making a note while it's fresh in my mind from PyCon US where the steering council solicited maintainer help:

It would also be helpful to evaluate the impact of [free threading](https://docs.python.org/3.13/whatsnew/3.13.html#free-threaded-cpython) on Sopel. Of course we cannot do this until we can run on 3.13 _at all_, so it's follow-up work.

**Comment by dgw** [comment]
I don't think it _should_ break anything, but what we need to do depends on what kind of help they want. Benchmark comparisons of our package running on GIL-enabled and GIL-disabled builds?

For our own purposes, testing on no-GIL Python would be good to expose any operations that are technically thread-unsafe but blocked by the GIL in current CPython builds. (But _I think_ we already have `Lock`s and such around those places. Probably.)

**Comment by SnoopJ** [comment]
Mostly I think it's a good stress test for upstream's benefit just to see if any wheels fall off the wagon. I doubt they will, but it's easy enough to give it a go.

I guess it would be nice to do a perf comparison (the build with `--disable-gil` can still have the GIL turned back on, so it's only one build to juggle), but that does almost certainly require some thought about what we would even measure (and how). I'd say perf eval is well worth splitting into a separate ticket if we end up wanting to see what the impact is

**Comment by SnoopJ** [comment]
3.13 is [fully released](https://docs.python.org/3.13/whatsnew/3.13.html) now

**Comment by SnoopJ** [comment]
Test suite is green for me on 3.13. I did have to disable a few requirements in `dev-requirements.txt`, but that would only impact `mypy`.

```
1387 passed, 8 xfailed, 22 warnings in 76.40s (0:01:16)
```

21 of those warnings are caused by passing `count` by position in `dice`:

```
sopel/builtins/dice.py: 21 warnings
  /var/lib/gitea/data/gitea-repositories/snoopj/sopel.git/sopel/builtins/dice.py:272: DeprecationWarning: 'count' is passed as positional argument
    arg_str = re.sub(dice_regexp, "%s", arg_str, 0, re.IGNORECASE | re.VERBOSE)
```

The other one is a SQLAlchemy thing:

```
test/test_db.py::test_execute                                                                                                                                           
  /var/lib/gitea/data/gitea-repositories/snoopj/sopel.git/sopel/db.py:337: RemovedIn20Warning: Deprecated API features detected! These feature(s) are not compatible wit
h SQLAlchemy 2.0. To prevent incompatible upgrades prior to updating applications, ensure requirements files are pinned to "sqlalchemy<2.0". Set environment variable SQ
LALCHEMY_WARN_20=1 to show all deprecation warnings.  Set environment variable SQLALCHEMY_SILENCE_UBER_WARNING=1 to silence this message. (Background on SQLAlchemy 2.0 
at: https://sqlalche.me/e/b8d9)
    return self.engine.execute(*args, **kwargs)
```

**Comment by SnoopJ** [comment]
> You shouldn't have had a types-pkg-resources requirement to disable; it was removed in https://github.com/sopel-irc/sopel/pull/2614 (https://github.com/sopel-irc/sopel/commit/c04ac82016393feeb5d52477a01561c339e44bb8). Is your clone up to date with master?

Ah, indeed I had forgotten to pull.

The error with `sqlalchemy[mypy]` is actually resolved by this as well. Re-running the suite now to be sure nothing got stirred up in the commits I was missing.

**Comment by SnoopJ** [comment]
Yup, all good:

```
1387 passed, 8 xfailed, 22 warnings in 148.33s (0:02:28)
```

**Comment by SnoopJ** [comment]
Looks like the `sqlalchemy[mypy]` thing was fixed upstream in 1.4.53 https://docs.sqlalchemy.org/en/20/changelog/changelog_14.html#change-a623aa18a66956a979824d53d56b22a5

**Comment by dgw** [comment]
Hmm, https://docs.sqlalchemy.org/en/20/changelog/changelog_14.html#change-1.4.53-mypy in particular.

It wouldn't be ideal to put #2492 in a minor release, since Sopel's DB API needs to change to accommodate some removals in SQLAlchemy itself. We're probably better off pinning `mypy<1.11` if we start running into problems, then prioritize SQLA 2.0 + bringing `mypy` current again soon after releasing Sopel 8.1.
